web/index.html

```diff
@@ -552,64 +601,147 @@ <h3>Used in</h3>
                     if (n === undefined || n <= 0) {
                         return "#####"
                     }
                     return n.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',')
                 },
-                loadChunk: async function (chunk) {
-                    return fetch(DATA_PREFIX + `chunks/chunk-${chunk}.json`)
-                        .then(response => {
-                            console.log("Loaded chunk", chunk);
-                            return response.json()
-                        })
-                        .then(data => {
-                            let [sep1, sep2, sep3] = [":3", ":2", ":1"]
-                            chunkData = Object.fromEntries(Object.entries(data).map(x => {
-                                let [from, to, hiddenFrom, hiddenTo] = x[1].split(sep1)
-                                if (from == "") {
-                                    from = []
-                                } else {
-                                    from = from.split(sep2).map(x => x.split(sep3))
-                                }
-                                if (to == "") {
-                                    to = []
-                                } else {
-                                    to = to.split(sep2).map(x => x.split(sep3))
-                                }
-                                if (hiddenFrom == "") {
-                                    hiddenFrom = []
-                                } else {
-                                    hiddenFrom = hiddenFrom.split(sep2).map(x => x.split(sep3))
-                                }
-                                if (hiddenTo == "") {
-                                    hiddenTo = []
+                calcCraftingTree: async function (element) {
+                    if (this.data.costs[element] == 1) {
+                        this.craftingTreeData = [];
+                        return;
+                    }
+                    this.craftingTreeData = null;
+                    let usedElements = new Set();
+                    let loopingElements = new Set();
+                    let result = [];
+                    let tree = {
+                        value: element
+                    };
+                    let craftingList = [tree];
+                    while (craftingList.length > 0) {
+                        let currentNode = craftingList.shift();
+                        let currentElement = currentNode.value;
+
+                        if (usedElements.has(currentElement)) {
+                            continue;
+                        }
+
+                        if (this.saveData.size == 0 || this.showFullRecipe) {
+                            currentNode.left = { value: this.data.craftingCache[currentElement][0] };
+                            currentNode.right = { value: this.data.craftingCache[currentElement][1] };
+                            usedElements.add(currentElement);
```

Comment:
If the save data is not loaded, or we are viewing the full recipe not, then we are always going to fall back to using the cached recipes as this will avoid ever needing to load additional chunks, which is still faster. It also is just the only method that makes sense once we have user data, as the additional sorting that is done to the from list with user data would make calculating the crafting recipe using the from list impossible without first resorting it.